### PR TITLE
JBPM-7906 - Audit Mode should apply to task log information

### PR DESCRIPTION
**Pull request by mswiderski**
depends on https://github.com/kiegroup/droolsjbpm-knowledge/pull/343

### CODE REVIEW

**Comment by tsurdilo** [comment]
@mswiderski looks good to me, only confusing part when looking was the priority numbers on calls to

protected void sendMessage(Object messageContent, Integer eventType, int priority)

Maybe using constants with some more meaningful names like HIGH_PRIORITY, LOW_PRIORITY etc would make it easier to understand. wdyt?

**Comment by mswiderski** [comment]
the priorities are from JMS api (0-9) and we could make them as constants but then how to name them when we use most of them ...

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/audit/StringCaseVariableIndexer.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        return true;
+    }
+
+    @Override
+    public List<CaseFileData> index(String name, Object variable) {
+        List<CaseFileData> indexed = new ArrayList<CaseFileData>();
+        
+        CaseFileDataLog caseVariable = new CaseFileDataLog();
+        caseVariable.setItemName(name);
+        caseVariable.setItemValue(variable == null ? "" : variable.toString());
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/audit/StringCaseVariableIndexer.java]: What about doing the same for `StringTaskVariableIndexer` here?
https://github.com/kiegroup/jbpm/blob/25fabbca948df52f69475ca54a8cf85fe09e4074/jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/variable/StringTaskVariableIndexer.java#L41-L50

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/audit/CaseInstanceAuditLoggerFactory.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            // look up connection factory and queue if given as property
+            if (properties.containsKey("jbpm.audit.jms.connection.factory.jndi")) {
+                ConnectionFactory connFactory = (ConnectionFactory) InitialContext.doLookup(
+                        (String)properties.get("jbpm.audit.jms.connection.factory.jndi")); 
+                logger.setConnectionFactory(connFactory);
+            }                
+            if (properties.containsKey("jbpm.audit.jms.queue.jndi")) {
+                Queue queue = (Queue) InitialContext.doLookup((String)properties.get("jbpm.audit.jms.queue.jndi")); 
+               logger.setQueue(queue);
+            }
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/audit/CaseInstanceAuditLoggerFactory.java]: yes, though not that likely

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/event/CaseConfigurationDeploymentListener.java [hunk 3]
```diff
@@ -145,4 +149,38 @@ public void onDeactivate(DeploymentEvent event) {
...
+            } else {                
+                return CaseInstanceAuditLoggerFactory.newJPAInstance(EntityManagerFactoryManager.get().getOrCreate(descriptor.getAuditPersistenceUnit()));
+            }
+        } else if (descriptor.getAuditMode() == AuditMode.JMS) {
+            
+            try {
+                Properties properties = new Properties();
+                InputStream input = manager.getEnvironment().getClassLoader().getResourceAsStream("/jbpm.audit.jms.properties");
+                if (input == null) {
+                    input = manager.getEnvironment().getClassLoader().getResourceAsStream("jbpm.audit.jms.properties");
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/event/CaseConfigurationDeploymentListener.java]: true, I have done it like this to handle kjars and unit tests - see DefaultRegisterableItemsFactory

File: jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/impl/model/TaskEventImpl.java [hunk 4]
```diff
@@ -47,7 +44,9 @@
 @Entity
 @Table(name = "TaskEvent")
 @SequenceGenerator(name = "taskEventIdSeq", sequenceName = "TASK_EVENT_ID_SEQ")
-public class TaskEventImpl implements TaskEvent, Externalizable {
+public class TaskEventImpl implements TaskEvent, Serializable {
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/impl/model/TaskEventImpl.java]: it was not suppose to be externalizable from the beginning ... and it was failing on JMS layer :)

File: jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/jms/AsyncTaskLifeCycleEventProducer.java [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+        
+        return auditTaskImpl;
+    }
+    
+    /*
+     * Not implemented task life cycle methods
+     */
+    
+    @Override
+    public void beforeTaskActivatedEvent(TaskEvent event) {
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/jms/AsyncTaskLifeCycleEventProducer.java]: these are simply not used for our auditing purposes

File: jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/jms/AsyncTaskLifeCycleEventProducer.java [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+    public void afterTaskAssignmentsAddedEvent(TaskEvent event, AssignmentType type, List<OrganizationalEntity> entities) {                
+        assignmentsUpadted(event, type, entities, "] have been added");    
+    }
+
+    @Override
+    public void afterTaskAssignmentsRemovedEvent(TaskEvent event, AssignmentType type, List<OrganizationalEntity> entities) {
+        assignmentsUpadted(event, type, entities, "] have been removed");
+    }
+    
+    protected void assignmentsUpadted(TaskEvent event, AssignmentType type, List<OrganizationalEntity> entities, String messageSufix) {
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/jms/AsyncTaskLifeCycleEventProducer.java]: fixed

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CompositeAsyncAuditLogReceiver.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    
+    /*
+     * Helper methods
+     */
+    
+    @SuppressWarnings("unchecked")
+    protected MessageListener createCaseEventReceiver() {
+        try {
+            Class<MessageListener> caseEventReceiverClass = (Class<MessageListener>) Class.forName("org.jbpm.casemgmt.impl.jms.AsyncCaseInstanceAuditEventReceiver");
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CompositeAsyncAuditLogReceiver.java]: due to the fact that case-mgmt-impl might not be on class path

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/audit/StringCaseVariableIndexer.java (hunk 1 above)
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/audit/StringCaseVariableIndexer.java]: fixed

File: jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/jms/AsyncTaskLifeCycleEventProducer.java [hunk 8]
```diff
@@ -0,0 +1,757 @@
...
+                                                     task.getTaskData().getWorkItemId(), 
+                                                     userId, message);
+        AuditTaskImpl auditTaskImpl = createAuditTask(task, event.getEventDate());
+        auditTaskImpl.setLastModificationDate(event.getEventDate());
+        
+        sendMessage(new AuditTaskData(auditTaskImpl, Collections.singletonList(taskEvent), null, taskVariables), 2);
+    }
+
+    @Override
+    public void afterTaskInputVariableChangedEvent(TaskEvent event, Map<String, Object> variables) {
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-human-task/jbpm-human-task-audit/src/main/java/org/jbpm/services/task/audit/jms/AsyncTaskLifeCycleEventProducer.java]: there is no need for an event as it is done as part of task creation so no need for another event in db

fixed the inconsistency with updating audit task here

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Pull request merged**